// element-plus@2.3.12/es/components/checkbox/src/checkbox-group2.mjs
import { defineComponent as n, computed as d, provide as f, toRefs as b, watch as h, openBlock as k, createBlock as x, resolveDynamicComponent as v, unref as r, normalizeClass as _, withCtx as g, renderSlot as I, nextTick as C } from "vue";
import { checkboxGroupProps as E, checkboxGroupEmits as y } from "./checkbox-group.mjs";
import { checkboxGroupContextKey as G } from "./constants.mjs";
import w from "../../../_virtual/plugin-vue_export-helper.mjs";
import { useNamespace as V } from "../../../hooks/use-namespace/index.mjs";
import { useFormItem as B, useFormItemInputId as D } from "../../form/src/hooks/use-form-item.mjs";
import { UPDATE_MODEL_EVENT as F } from "../../../constants/event.mjs";
import { debugWarn as T } from "../../../utils/error.mjs";
import z from "../../../../../../../registry.npmmirror.com_lodash-es@4.17.21/node_modules/lodash-es/pick.mjs";
const L = n({
  name: "ElCheckboxGroup"
}), N = /* @__PURE__ */ n({
  ...L,
  props: E,
  emits: y,
  setup(s, { emit: a }) {
    const o = s, i = V("checkbox"), { formItem: t } = B(), { inputId: p, isLabeledByFormItem: l } = D(o, {
      formItemContext: t
    }), c = async (e) => {
      a(F, e), await C(), a("change", e);
    }, u = d({
      get() {
        return o.modelValue;
      },
      set(e) {
        c(e);
      }
    });
    return f(G, {
      ...z(b(o), [
        "size",
        "min",
        "max",
        "disabled",
        "validateEvent",
        "fill",
        "textColor"
      ]),
      modelValue: u,
      changeEvent: c
    }), h(() => o.modelValue, () => {
      o.validateEvent && (t == null || t.validate("change").catch((e) => T(e)));
    }), (e, P) => {
      var m;
      return k(), x(v(e.tag), {
        id: r(p),
        class: _(r(i).b("group")),
        role: "group",
        "aria-label": r(l) ? void 0 : e.label || "checkbox-group",
        "aria-labelledby": r(l) ? (m = r(t)) == null ? void 0 : m.labelId : void 0
      }, {
        default: g(() => [
          I(e.$slots, "default")
        ]),
        _: 3
      }, 8, ["id", "class", "aria-label", "aria-labelledby"]);
    };
  }
});
var j = /* @__PURE__ */ w(N, [["__file", "/home/runner/work/element-plus/element-plus/packages/components/checkbox/src/checkbox-group.vue"]]);
export {
  j as default
};
