// element-plus@2.3.12/es/components/collapse-transition/src/collapse-transition.mjs
import { defineComponent as e, openBlock as s, createBlock as n, Transition as i, mergeProps as r, unref as l, toHandlers as p, withCtx as m, renderSlot as g } from "vue";
import f from "../../../_virtual/plugin-vue_export-helper.mjs";
import { useNamespace as c } from "../../../hooks/use-namespace/index.mjs";
const y = e({
  name: "ElCollapseTransition"
}), h = /* @__PURE__ */ e({
  ...y,
  setup(v) {
    const a = c("collapse-transition"), o = (t) => {
      t.style.maxHeight = "", t.style.overflow = t.dataset.oldOverflow, t.style.paddingTop = t.dataset.oldPaddingTop, t.style.paddingBottom = t.dataset.oldPaddingBottom;
    }, d = {
      beforeEnter(t) {
        t.dataset || (t.dataset = {}), t.dataset.oldPaddingTop = t.style.paddingTop, t.dataset.oldPaddingBottom = t.style.paddingBottom, t.style.maxHeight = 0, t.style.paddingTop = 0, t.style.paddingBottom = 0;
      },
      enter(t) {
        t.dataset.oldOverflow = t.style.overflow, t.scrollHeight !== 0 ? t.style.maxHeight = `${t.scrollHeight}px` : t.style.maxHeight = 0, t.style.paddingTop = t.dataset.oldPaddingTop, t.style.paddingBottom = t.dataset.oldPaddingBottom, t.style.overflow = "hidden";
      },
      afterEnter(t) {
        t.style.maxHeight = "", t.style.overflow = t.dataset.oldOverflow;
      },
      enterCancelled(t) {
        o(t);
      },
      beforeLeave(t) {
        t.dataset || (t.dataset = {}), t.dataset.oldPaddingTop = t.style.paddingTop, t.dataset.oldPaddingBottom = t.style.paddingBottom, t.dataset.oldOverflow = t.style.overflow, t.style.maxHeight = `${t.scrollHeight}px`, t.style.overflow = "hidden";
      },
      leave(t) {
        t.scrollHeight !== 0 && (t.style.maxHeight = 0, t.style.paddingTop = 0, t.style.paddingBottom = 0);
      },
      afterLeave(t) {
        o(t);
      },
      leaveCancelled(t) {
        o(t);
      }
    };
    return (t, _) => (s(), n(i, r({
      name: l(a).b()
    }, p(d)), {
      default: m(() => [
        g(t.$slots, "default")
      ]),
      _: 3
    }, 16, ["name"]));
  }
});
var x = /* @__PURE__ */ f(h, [["__file", "/home/runner/work/element-plus/element-plus/packages/components/collapse-transition/src/collapse-transition.vue"]]);
export {
  x as default
};
